File: src/net/polyengine/Entity.java
package net.polyengine;

import com.sun.istack.internal.NotNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class Entity {

	private boolean registered = false;
	private boolean binned = false;
	private boolean destroyed = false;

	final List<Component> mComponents = new ArrayList<>();

	public final World world;
	public final List<Component> components = Collections.unmodifiableList(mComponents);

	Entity(World world) {
		this.world = world;
	}



	public boolean isRegistered() {
		return registered;
	}

	public boolean isBinned() {
		return binned;
	}

	public boolean isDestroyed() {
		return destroyed;
	}



	void register() {
		assert(!registered && !binned && !destroyed);
		assert(world.isRegistered() && !world.isDestroyed());

		world.mEntities.add(this);
		registered = true;
	}

	public void destroy() {
		// TODO: Error handling
		if (!registered) throw new RuntimeException("The entity hasn't been registered.");
		if (destroyed) throw new RuntimeException("The entity has already been destroyed.");

		assert(world.isRegistered() && !world.isDestroyed());

		if (!binned) {
			Engine.destroyingEntities.add(this);
			for (Component component : components) component.destroy();
			binned = true;
		}
	}



	void performDestroy() {
		assert(registered && binned && !destroyed);

		world.mEntities.remove(this);
		mComponents.clear();
		registered = false;
		binned = false;
		destroyed = true;
	}



	public <T extends Component> T addComponent(@NotNull Class<T> componentClass) {
		// TODO: Error handling
		if (!registered) throw new RuntimeException("The entity hasn't been registered.");
		if (destroyed) throw new RuntimeException("The entity has been destroyed.");

		T component = Component.newComponent(componentClass, this);
		component.register();
		return component;
	}



	public <T extends Component> T getComponent(@NotNull Class<T> componentClass) {
		for (Component component : components) {
			if (componentClass.isInstance(component)) {
				return componentClass.cast(component);
			}
		}
		return null;
	}

	public <T extends Component> List<T> getComponents(@NotNull Class<T> componentClass) {
		List<T> components = new ArrayList<>();
		for (Component component : this.components) {
			if (componentClass.isInstance(component)) {
				components.add(componentClass.cast(component));
			}
		}
		return components;
	}
}

File: src/net/polyengine/Entity.kt
package net.polyengine

import net.polyengine.listeners.EntityListener
import java.util.*
import kotlin.reflect.KClass
import kotlin.reflect.full.cast

@Suppress("unused", "MemberVisibilityCanBePrivate")
class Entity internal constructor(val world: World, attributeClasses: List<KClass<out Attribute>>) {

	val engine: Engine = world.engine
	val attributes: List<Attribute>

	var destroyed = false
		private set

	private var destroyRequested = false

	init {
		val mutableAttributes: MutableList<Attribute> = LinkedList()
		attributeClasses.forEach { mutableAttributes.add(Attribute.new(it, this@Entity)) }

		attributes = Collections.unmodifiableList(mutableAttributes)
	}

	// Events

	private val createEvent = object : Event {

		override fun handle() {
			if (!world.destroyed) {
				world.addEntity(this@Entity)
				attributes.forEach { it.init() }

				engine.extensions.forEach {
					if (it is EntityListener) it.onEntityCreated(this@Entity)
				}
			}
		}
	}

	internal val destroyEvent = object : Event {

		override fun handle() {
			if (!world.destroyed) {
				destroyed = true
				attributes.forEach { it.term() }
				world.removeEntity(this@Entity)

				engine.extensions.forEach {
					if (it is EntityListener) it.onEntityDestroyed(this@Entity)
				}
			}
		}
	}

	// Actions

	init {
		engine.sendEvent(createEvent)
	}

	fun destroy() {
		if (destroyed) throw IllegalStateException("The entity has been destroyed")
		if (!destroyRequested) {
			destroyRequested = true
			engine.sendEvent(destroyEvent)
		}
	}

	// Queries

	fun <T : Attribute> getAttribute(attributeClass: KClass<T>): T? {
		for (attribute in attributes) {
			if (attributeClass.isInstance(attribute)) {
				return attributeClass.cast(attribute)
			}
		}
		return null
	}

	fun <T : Attribute> getAttribute(attributeClass: Class<T>): T? {
		return getAttribute(attributeClass.kotlin)
	}

	fun <T : Attribute> getAttributes(attributeClass: KClass<T>): List<T> {
		val attributes = LinkedList<T>()
		for (attribute in this.attributes) {
			if (attributeClass.isInstance(attribute)) {
				attributes.add(attributeClass.cast(attribute))
			}
		}
		return Collections.unmodifiableList(attributes)
	}

	fun <T : Attribute> getAttributes(attributeClass: Class<T>): List<T> {
		return getAttributes(attributeClass.kotlin)
	}
}